master/internal/core_trial.go

```diff
@@ -40,6 +45,123 @@ func echoCanGetTrial(c echo.Context, m *Master, trialID string) error {
+	}
+
+	var trialIDs []int
+	for _, id := range strings.Split(in, ",") {
+		if err := echoCanGetTrial(c, m, id); err != nil {
+			return nil, err
+		}
+
+		trialID, err := strconv.Atoi(id)
+		if err != nil {
+			return nil, echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf(
+				"could not parse trialIds parameter expected comma separated ints got %s", in))
+		}
+		trialIDs = append(trialIDs, trialID)
+	}
+
+	return trialIDs, nil
+}
+
+//	@Summary	Stream one or more trial's training metrics.
+//	@Tags		Trials
+//	@ID			stream-training-metrics
+//	@Produce	application/x-ndjson
+//	@Param		trialIds	query	string	true	"Comma delimited trial IDs"
+//	@Success	200					{}		string	"A new line delimated JSON stream with the following fields trialId,endTime,metrics,totalBatches,trialRunId,id"
+//	@Router		/trials/metrics/training [get]
+//
+// nolint:lll
+func (m *Master) streamTrainingMetrics(c echo.Context) error {
+	trialIDs, err := m.parseAndRBACTrialIDs(c, c.QueryParam("trialIds"))
+	if err != nil {
+		return err
+	}
+
+	err = streamQueryAsNDJSON(c, `SELECT jsonb_build_object(
+		'trialId', trial_id,
+		'endTime', end_time,
+		'metrics', metrics,
+		'totalBatches', total_batches,
+		'trialRunId', trial_run_id,
+		'id', id
+	) FROM steps WHERE trial_id IN (?) ORDER BY trial_id, trial_run_id, total_batches`,
+		bun.In(trialIDs))
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError,
+			errors.Wrap(err, "error streaming training metrics"))
+	}
+	return nil
+}
+
+//	@Summary	Stream one or more trial's validation metrics.
+//	@Tags		Trials
+//	@ID			stream-validation-metrics
+//	@Produce	application/x-ndjson
+//	@Param		trialIds	query	string	true	"Comma delimited trial IDs"
+//	@Success	200					{}		string	"A new line delimated JSON stream with the following fields trialId,endTime,metrics,totalBatches,trialRunId,id"
+//	@Router		/trials/metrics/validation [get]
+//
+// nolint:lll
+func (m *Master) streamValidationMetrics(c echo.Context) error {
+	trialIDs, err := m.parseAndRBACTrialIDs(c, c.QueryParam("trialIds"))
+	if err != nil {
+		return err
+	}
+
+	err = streamQueryAsNDJSON(c, `SELECT jsonb_build_object(
+		'trialId', trial_id,
+		'endTime', end_time,
+		'metrics', metrics,
+		'totalBatches', total_batches,
+		'trialRunId', trial_run_id,
+		'id', id
+	) FROM validations WHERE trial_id IN (?) ORDER BY trial_id, trial_run_id, total_batches`,
+		bun.In(trialIDs))
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError,
+			errors.Wrap(err, "error streaming training metrics"))
+	}
+	return nil
+}
+
+func streamQueryAsNDJSON(c echo.Context, query string, args ...any) error {
+	rows, err := db.Bun().QueryContext(c.Request().Context(), query, args...)
+	if err != nil {
+		return errors.Wrap(err, "error running query to stream")
+	}
+	defer rows.Close()
+
+	c.Response().Header().Set(echo.HeaderContentType, "application/x-ndjson")
+	c.Response().WriteHeader(http.StatusOK)
+
+	for rows.Next() {
```

Comment:
we have avoided exactly this idea in the past because it pins a db connection for the duration of the streaming request (we shockingly had this cause a deadlock). so, just to note, this could grind the cluster to a halt if you hit https://github.com/determined-ai/determined/blob/f429b37e3d686745d0c74f302f5cbd57b16df083/master/internal/db/setup.go#L12

Reply:
back to the drawing board.

Reply:
Shouldn't be too much of an issue to do multiple queries with some kind of keyset pagination.